lib/puppet/file_serving/metadata.rb

```diff
@@ -138,7 +138,7 @@ def initialize(path,data={})
     @source      = data.delete('source')
     @content_uri = data.delete('content_uri')
 
-    super(path, data.map { |k, v| [k.to_sym, v] }.to_h)
+    super(path, **data.map { |k, v| [k.to_sym, v] }.to_h)
```

Comment:
Since the parent `Base` class accepts keyword arguments, maybe it would be better to pass those explicitly here, like `links = data.fetch(:links, nil)` and then `super(path, links: links, ...)`? That said, that sounds like different issue, as currently, if you pass an option in `data` that isn't a keyword argument that `Base` accepts, it will raise an error. In general, I think we want to ignore fields we don't understand, such as if an older agent deserializes a metadata object from a newer puppetserver.

Reply:
I *think* this is what you're suggesting I do instead?
```
diff --git a/lib/puppet/file_serving/metadata.rb b/lib/puppet/file_serving/metadata.rb
index e30ecf970f..299707408e 100644
--- a/lib/puppet/file_serving/metadata.rb
+++ b/lib/puppet/file_serving/metadata.rb
@@ -138,7 +138,9 @@ class Puppet::FileServing::Metadata < Puppet::FileServing::Base
     @source      = data.delete('source')
     @content_uri = data.delete('content_uri')

-    super(path, **data.map { |k, v| [k.to_sym, v] }.to_h)
+    links = data.fetch('links', nil)
+    relative_path = data.fetch('relative_path', nil)
+    super(path, links: links, relative_path: relative_path, source: @source)
   end
```

This causes a bunch of errors, unfortunately
```
  1) Puppet::FileServing::Metadata  when pointing to a link when digest_algorithm is md5 when links are followed should not store the destination of the link in :destination if links are :follow
     Failure/Error: File.lstat(path)

     Errno::ENOENT:
       No such file or directory @ rb_file_s_lstat - /base/path/my/file
     # ./lib/puppet/file_system/file_impl.rb:147:in `lstat'
     # ./lib/puppet/file_system/file_impl.rb:147:in `lstat'
     # ./lib/puppet/file_system.rb:314:in `lstat'
     # ./lib/puppet/file_serving/base.rb:72:in `stat'
     # ./lib/puppet/file_serving/metadata.rb:89:in `collect_stat'
     # ./lib/puppet/file_serving/metadata.rb:104:in `collect'
     # ./spec/unit/file_serving/metadata_spec.rb:541:in `block (4 levels) in <top (required)>'
     # ./spec/spec_helper.rb:180:in `block (2 levels) in <top (required)>'
```

Reply:
I'm starting to think that failure is because of a mismatch of expectations. It looks like we have some inconsistency here between if we're using strings or symbols, which is frustrating...